File: src/core/clauses.ts
/**
 * Assemble a phonemizer's text() output from its tokenizer. Collapses the identical `out`/`pending` + `emit`
 * closure + final-flush skeleton that every language's text() repeats: phonemized tokens are space-joined, and a
 * clause-punctuation mark becomes a PENDING pause attached BEFORE the next emitted token (never trailing an empty
 * output, never doubled). Only the per-token dispatch — which regex group is a word / number / punctuation, and
 * how each is phonemized — varies, so the caller supplies a `handle(match, sink)` over `input.matchAll(token)`.
 *
 *   text(input) {
 *     return assembleClauses(input, TOKEN, (m, sink) => {
 *       if (m[1]) sink.emit(phonemizeWord(m[1]));
 *       else if (m[2]) for (const w of numberWords(m[2])) sink.emit(phonemizeWord(w));
 *       else if (m[3]) { const mk = CLAUSE_MARK[m[3]]; if (mk) sink.pause(mk); }
 *     });
 *   }
 */
import { getDefaultForeign, readForeignRun } from "./foreign.ts";

export interface ClauseSink {
    /** Append a phonemized token, space-joined; flushes any pending pause before it. Empty strings are ignored. */
    emit(ipa: string): void;
    /** Set a pending clause pause (rendered before the next emitted token). No-op while the output is still empty. */
    pause(mark: string): void;
}

/**
 * The clause-assembly state machine, independent of how tokens are iterated. `sink` drives it; `finish()`
 * flushes any trailing pause and returns the assembled string. Use this directly when the tokenization isn't a
 * single regex matchAll (mandarin scans code-point Han/Latin runs); use assembleClauses otherwise.
 */
export function clauseSink(): { sink: ClauseSink; finish: () => string } {
    let out = "";
    let pending: string | null = null;
    const sink: ClauseSink = {
        emit(ipa) {
            if (ipa === "") return;
            if (out === "") out = ipa;
            else if (pending !== null) {
                out += ` ${pending} ${ipa}`;
                pending = null;
            } else out += ` ${ipa}`;
        },
        pause(mark) {
            if (out !== "") pending = mark;
        },
    };
    return {
        sink,
        finish() {
            if (pending !== null && out !== "") out += ` ${pending}`;
            return out;
        },
    };
}

/** A run of Latin-script text (with its combining marks, apostrophes and internal hyphens). */
const LATIN_RUN = /\p{Script=Latin}[\p{Script=Latin}\p{M}'’-]*/gu;
/**
 * A run of letters in ANY script, kept together with its combining marks and internal apostrophes.
 *
 * EXPORTED because two engines cannot use `assembleClauses` and hand-roll their own gap pass —
 * english.ts (a two-phase tagger pipeline) and french.ts (liaison looks one word ahead). One definition,
 * three call sites.
 *
 * ⚠ DELIBERATELY LETTERS ONLY, not `\p{Nd}`. A digit is script-marked but language-NEUTRAL in value, so
 * `٢٠٢٤` in English text is 2024 and wants an English reading, not an Arabic one. The registry folds native
 * digits to ASCII for every language before any engine sees them, which is a better answer than routing them
 * by script.
 *
 * ⚠ A TRAILING SUPERSCRIPT TRAVELS WITH THE RUN, because it belongs to the foreign expression and not to the
 * host. In `E = mc²` the `²` is `No`, not `\p{L}`, so a letters-only run ends at `mc` and the exponent is left
 * in the gap and DROPPED. English reads `mc²` perfectly well once handed the whole thing, so the fix is to
 * stop cutting the expression in half rather than to invent a host-language reading for it.
 * Only TRAILING, and only the superscript digits: a superscript cannot begin a word, so this cannot start a
 * run that would not otherwise exist, and it cannot swallow a following host-language character.
 */
export const FOREIGN_RUN = /[\p{L}\p{M}][\p{L}\p{M}'’-]*[\u2070\u00b9\u00b2\u00b3\u2074-\u2079]*/gu;

/**
 * Emit the FOREIGN runs inside text the engine's own tokenizer did not claim.
 *
 * A run in ANY script is a candidate: the script router is asked first, and only if it declines does the
 * Latin-to-English fallback apply. Everything else in a gap — stray punctuation, whitespace — stays dropped
 * (see core/foreign.ts for why an unclaimed gap loses embedded text entirely).
 */
export function emitUnclaimed(gap: string, sink: ClauseSink): void {
    for (const m of gap.matchAll(FOREIGN_RUN)) {
        const run = m[0];
        // The script router first — it knows which language should read this run. It declines for an
        // unknown script, and for a target equal to the host (which would recurse).
        const routed = readForeignRun(run);
        if (routed !== undefined) {
            if (routed !== "") sink.emit(routed);
            continue;
        }
        // Fallback: the Latin-to-English path, for wherever the router has nothing to say.
        if (!/\p{Script=Latin}/u.test(run)) continue;
        const foreign = getDefaultForeign();
        if (foreign !== undefined) sink.emit(foreign(run));
    }
}

export function assembleClauses(
    input: string,
    token: RegExp,
    handle: (match: RegExpMatchArray, sink: ClauseSink) => void,
): string {
    const { sink, finish } = clauseSink();
    let cursor = 0;
    for (const m of input.matchAll(token)) {
        const at = m.index ?? cursor;
        // Text between the previous match and this one is text the tokenizer declined; an engine whose
        // TOKEN already has a Latin group claims it here and leaves no gap, so this is a no-op for the
        // engines that were already correct.
        if (at > cursor) emitUnclaimed(input.slice(cursor, at), sink);
        handle(m, sink);
        cursor = at + m[0].length;
    }
    if (cursor < input.length) emitUnclaimed(input.slice(cursor), sink);
    return finish();
}

File: csharp/Vernacula.Phonemizer/Core/Clauses.cs
/**
 * Assemble a phonemizer's text() output from its tokenizer.
 * Ported from src/core/clauses.ts — see that file for the corpus evidence.
 */
using System.Text.RegularExpressions;

namespace Vernacula.Phonemizer.Core;

public interface ClauseSink
{
    /**
     * Append a phonemized token, space-joined; flushes any pending pause before it. Empty strings are
     * ignored.
     */
    void Emit(string ipa);
    /**
     * Set a pending clause pause (rendered before the next emitted token). No-op while the output is still
     * empty.
     */
    void Pause(string mark);
}

public static class Clauses
{
    /** The clause-assembly state machine, independent of how tokens are iterated. */
    public static (ClauseSink sink, Func<string> finish) ClauseSink()
    {
        var impl = new SinkImpl();
        return (impl, impl.Finish);
    }

    private sealed class SinkImpl : ClauseSink
    {
        private string @out = "";
        private string? pending = null;

        public void Emit(string ipa)
        {
            if (ipa == "") return;
            // ⚠ #1150 STAGE 3: THE OFFSET IS COMPUTED BEFORE THE APPEND, and the separator is part of the
            // arithmetic rather than an afterthought — `at` is where THIS emission starts in the assembled
            // reading, which is the only thing that makes a token's contribution locatable in the output.
            int at;
            if (@out == "")
            {
                at = 0;
                @out = ipa;
            }
            else if (pending != null)
            {
                at = @out.Length + pending.Length + 2; // " " + mark + " "
                @out += $" {pending} {ipa}";
                pending = null;
            }
            else
            {
                at = @out.Length + 1; // " "
                @out += $" {ipa}";
            }
            Trace.NoteEmit(ipa, at);
        }

        public void Pause(string mark)
        {
            if (@out != "") pending = mark;
        }

        public string Finish()
        {
            if (pending != null && @out != "") @out += $" {pending}";
            // ⚠ THE ASSEMBLED STRING IS NOT NECESSARILY THE READING. Eight engines rewrite it afterwards; the
            // recorder compares this against the final IPA and withholds the output spans when they differ,
            // rather than reporting offsets into a string that no longer exists.
            Trace.NoteAssembled(@out);
            return @out;
        }
    }

    /** A run of Latin-script text (with its combining marks, apostrophes and internal hyphens). */
    // C# PORT NOTE: module-private and UNREFERENCED in the TS source too; kept `internal` (a private
    // field that is never read would draw CS0414).
    internal static readonly JsRe LATIN_RUN = JsRegex.Compile(@"\p{Script=Latin}[\p{Script=Latin}\p{M}'’-]*", "gu");
    /**
     * A run of letters in ANY script, kept together with its combining marks and internal apostrophes.
     *
     * ⚠ DELIBERATELY LETTERS ONLY, not `\p{Nd}`. A digit is script-marked but language-NEUTRAL in value, so
     * `٢٠٢٤` in English text is 2024 and wants an English reading; the registry folds native digits to ASCII
     * before any engine sees them, which is a better answer than routing them by script.
     *
     * ⚠ A TRAILING SUPERSCRIPT TRAVELS WITH THE RUN, because it belongs to the foreign expression and not to
     * the host: a letters-only run ends `E = mc²` at `mc` and the exponent is left in the gap and DROPPED.
     * Only trailing, and only the superscript digits — a superscript cannot begin a word, so this can never
     * start a run that would not otherwise exist.
     */
    public static readonly JsRe FOREIGN_RUN =
        JsRegex.Compile(@"[\p{L}\p{M}][\p{L}\p{M}'’-]*[\u2070\u00b9\u00b2\u00b3\u2074-\u2079]*", "gu");

    private static readonly JsRe LatinOne = JsRegex.Compile(@"\p{Script=Latin}", "u");

    /** Emit the FOREIGN runs inside text the engine's own tokenizer did not claim. */
    public static void EmitUnclaimed(string gap, ClauseSink sink, int baseOffset = 0)
    {
        foreach (Match m in FOREIGN_RUN.Matches(gap))
        {
            var run = m.Value;
            // ⚠ AN UNCLAIMED RUN IS STILL A TOKEN (#1150): its reading reaches the output, so leaving it out
            // would attribute real IPA to no token. `baseOffset` is the gap's offset in the caller's string.
            Trace.BeginToken(baseOffset + m.Index, baseOffset + m.Index + run.Length, run);
            var routed = Foreign.ReadForeignRun(run);
            if (routed != null)
            {
                if (routed != "") sink.Emit(routed);
                Trace.EndToken();
                continue;
            }
            if (!LatinOne.IsMatch(run)) { Trace.EndToken(); continue; }
            var foreign = Foreign.GetDefaultForeign();
            if (foreign != null) sink.Emit(foreign(run));
            Trace.EndToken();
        }
    }

    public static string AssembleClauses(
        string input,
        JsRe token,
        Action<Match, ClauseSink> handle)
    {
        var (sink, finish) = ClauseSink();
        // The trace is derived here and nowhere else (#1150): `at`/`cursor` were already computed and
        // discarded, and a token's span is that arithmetic kept.
        Trace.EnterEngine(input);
        try
        {
            var cursor = 0;
            foreach (Match m in JsRegex.MatchAll(token, input))
            {
                var at = m.Index; // TS `m.index ?? cursor` — a .NET Match always carries its index
                if (at > cursor) EmitUnclaimed(input[cursor..at], sink, cursor);
                Trace.BeginToken(at, at + m.Value.Length, m.Value);
                handle(m, sink);
                Trace.EndToken();
                cursor = at + m.Value.Length;
            }
            if (cursor < input.Length) EmitUnclaimed(input[cursor..], sink, cursor);
            return finish();
        }
        finally { Trace.ExitEngine(); }
    }
}
